tests/playwright-test/only-changed.spec.ts

```diff
@@ -0,0 +1,267 @@
+
+test('should detect untracked files', async ({ runInlineTest, setupRepository }) => {
+  await setupRepository();
+
+  const result = await runInlineTest({
+    'c.spec.ts': `
+      import { test, expect } from '@playwright/test';
+      test('fails', () => { expect(1).toBe(2); });
+    `
+  }, { 'only-changed': true });
+
+  expect(result.exitCode).toBe(1);
+  expect(result.failed).toBe(1);
+  expect(result.output).toContain('c.spec.ts');
+});
+
+
+test('should detect changed files', async ({ runInlineTest, setupRepository }) => {
+  await setupRepository();
+  const result = await runInlineTest({
+    'b.spec.ts': `
+        import { test, expect } from '@playwright/test';
+        test('fails', () => { expect(1).toBe(3); });
+      `,
+  }, { 'only-changed': true });
+
+  expect(result.exitCode).toBe(1);
+  expect(result.failed).toBe(1);
+  expect(result.output).toContain('b.spec.ts');
+});
+
+test('should diff based on base commit', async ({ runInlineTest, setupRepository, writeFiles }) => {
+  const git = await setupRepository();
+  await writeFiles({
+    'b.spec.ts': `
+        import { test, expect } from '@playwright/test';
+        test('fails', () => { expect(1).toBe(3); });
+      `,
+  });
+  git('commit -a -m update');
+  const result = await runInlineTest({}, { 'only-changed': `HEAD~1` });
+
+  expect(result.exitCode).toBe(1);
+  expect(result.failed).toBe(1);
+  expect(result.output).toContain('b.spec.ts');
+});
+
+test('should understand dependency structure', async ({ runInlineTest, setupRepository, writeFiles }) => {
+  await setupRepository();
+  await writeFiles({
+    'question.ts': `
+        export const question = "what is the answer to life the universe and everything";
+      `,
+  });
+  const result = await runInlineTest({}, { 'only-changed': true });
+
+  expect(result.exitCode).toBe(1);
+  expect(result.failed).toBe(2);
+  expect(result.output).toContain('a.spec.ts');
+  expect(result.output).toContain('b.spec.ts');
+});
+
+test('should support watch mode', async ({ setupRepository, writeFiles, runWatchTest }) => {
+  const git = await setupRepository();
+  await writeFiles({
+    'b.spec.ts': `
+        import { test, expect } from '@playwright/test';
+        test('fails', () => { expect(1).toBe(3); });
+      `,
+  });
+  git('commit -a -m update');
+
+  const testProcess = await runWatchTest({}, { 'only-changed': `HEAD~1` });
+  await testProcess.waitForOutput('Waiting for file changes.');
+  testProcess.clearOutput();
+  testProcess.write('r');
+
+  await testProcess.waitForOutput('b.spec.ts:3:13 › fails');
+  expect(testProcess.output).not.toContain('a.spec');
+});
+
+test('should throw nice error message if git doesnt work', async ({ setupRepository, runInlineTest }) => {
+  await setupRepository();
+  const result = await runInlineTest({}, { 'only-changed': `this-commit-does-not-exist` });
+
+  expect(result.exitCode).toBe(1);
+  expect(result.output).toContain('only works with Git repositories');
```

Comment:
Perhaps we should print the `git` output to make it easier to diagnose issues?

Reply:
yes! done in https://github.com/microsoft/playwright/pull/31727/commits/b6ef01324fa3dd0dc6bf70d348bb38d409ce7c55